### PR TITLE
Implementation for LSD v1

### PR DESCRIPTION
**Pull request by piobab**
v1 LSDs:
```
stTKN/USD = stTKN/TKN * TKN/USD where:
stTKN/TKN = min(stTKN/TKN TWAP, stTKN/TKN Redemption Rate)
```

For now keeping Lsd price source in Osmosis (Pyth too). I will do cleanup, package reorganization once we move to `mars-common` where Sturdy and Patrick work. I don't want to do too much in this PR and repo.

### CODE REVIEW

File: contracts/oracle/osmosis/src/stride.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+use cosmwasm_std::{to_binary, Addr, Decimal, QuerierWrapper, QueryRequest, StdResult, WasmQuery};
```
Comment (piobab) [inline contracts/oracle/osmosis/src/stride.rs:1]: This file maybe removed once Stride open source their codebase

File: contracts/oracle/base/src/contract.rs [hunk 2]
```diff
@@ -140,14 +146,14 @@ where
...
+        price_source: PU,
     ) -> ContractResult<Response> {
         self.owner.assert_owner(deps.storage, &sender_addr)?;
 
         validate_native_denom(&denom)?;
 
         let cfg = self.config.load(deps.storage)?;
-        price_source.validate(&deps.querier, &denom, &cfg.base_denom)?;
+        let price_source = price_source.validate(deps.as_ref(), &denom, &cfg.base_denom)?;
         self.price_sources.save(deps.storage, &denom, &price_source)?;
```
Comment (grod220) [inline contracts/oracle/base/src/contract.rs:157]: It seems like a good amount of extra checked vs unchecked code in order to support this pattern of validation. Is this worth it?

File: contracts/oracle/osmosis/src/price_source.rs [hunk 3]
```diff
@@ -452,6 +602,66 @@ impl OsmosisPriceSource {
...
+    fn query_lsd_price(
+        deps: &Deps,
+        env: &Env,
+        denom: &str,
+        transitive_denom: &str,
+        base_denom: &str,
+        geometric_twap: GeometricTwap,
+        redemption_rate: RedemptionRate<Addr>,
+        price_sources: &Map<&str, OsmosisPriceSourceChecked>,
+        pyth_config: &PythConfig,
```
Comment (piobab) [inline contracts/oracle/osmosis/src/price_source.rs:620]: I am going to remove pyth config from global config and keep it in Pyth related price sources.

File: contracts/oracle/osmosis/src/price_source.rs [hunk 4]
```diff
@@ -525,30 +735,30 @@ mod tests {
 
     #[test]
     fn display_fixed_price_source() {
-        let ps = OsmosisPriceSource::Fixed {
+        let ps = OsmosisPriceSourceChecked::Fixed {
```
Comment (piobab) [inline contracts/oracle/osmosis/src/price_source.rs:738]: I will move these test cases to separate file to make `price_source.rs` smaller

**Review by grod220 — APPROVED** [review]
Nice job! I think it's better to get this over to mars-common sooner rather than later if this overlaps with any of Sturdy/Patrick's work.

File: contracts/oracle/osmosis/src/price_source.rs [hunk 5]
```diff
@@ -159,9 +159,73 @@ pub enum OsmosisPriceSource {
...
+        /// Params to query geometric TWAP price
+        geometric_twap: GeometricTwap,
+
+        /// Params to query redemption rate
+        redemption_rate: RedemptionRate<T>,
+    },
 }
 
-impl fmt::Display for OsmosisPriceSource {
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
```
Comment (grod220) [inline contracts/oracle/osmosis/src/price_source.rs:190]: Maybe we can start using `#[cw_serde]` here now?

File: contracts/oracle/base/src/contract.rs (hunk 2 above)
Comment (piobab) [reply contracts/oracle/base/src/contract.rs:157]: We will need this in Patrick/Sturdy works too (they already included it). It adds extra struct creation so maybe not that big cost.. but it clear separation.

**Comment by piobab** [comment]
> Nice job! I think it's better to get this over to mars-common sooner rather than later if this overlaps with any of Sturdy/Patrick's work.

Yeah, agreed. That's the next step to include in `mars-common`. Once included I will do package reorganization and small cleanup.

File: contracts/oracle/osmosis/src/price_source.rs (hunk 5 above)
Comment (larry0x) [reply contracts/oracle/osmosis/src/price_source.rs:190]: Agree

File: contracts/oracle/osmosis/src/price_source.rs [hunk 6]
```diff
@@ -159,9 +159,73 @@ pub enum OsmosisPriceSource {
...
+
+impl From<&GeometricTwap> for GeometricTwap {
+    fn from(value: &GeometricTwap) -> Self {
+        Self {
+            pool_id: value.pool_id,
+            window_size: value.window_size,
+            downtime_detector: value.downtime_detector.clone(),
+        }
+    }
+}
```
Comment (larry0x) [inline contracts/oracle/osmosis/src/price_source.rs:212]: Isn't this just cloning? Can't you just use `clone()`

File: contracts/oracle/osmosis/src/price_source.rs [hunk 7]
```diff
@@ -159,9 +159,73 @@ pub enum OsmosisPriceSource {
...
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
```
Comment (larry0x) [inline contracts/oracle/osmosis/src/price_source.rs:214]: `#[cw_serde]`

**Review by larry0x — APPROVED** [review]

File: contracts/oracle/osmosis/src/price_source.rs (hunk 5 above)
Comment (piobab) [reply contracts/oracle/osmosis/src/price_source.rs:190]: I would like to but it screams that OsmosisPriceSource doesn't implement these traits:
`Serialize + DeserializeOwned + Clone + Debug + PartialEq + JsonSchema`

File: contracts/oracle/osmosis/src/price_source.rs (hunk 6 above)
Comment (piobab) [reply contracts/oracle/osmosis/src/price_source.rs:212]: Yes it is. I don't remember why I added it then

**Pull request merged**